# Анализ базы данных посредством SQL

**Цель исследования** — проанализировать базу данных сервиса для чтения книг по подписке, а именно:
1. определить количество книг выпущенных после 1 января 2000 года
2. определить количество обзоров и среднюю оценку для каждой книги
3. определить издательство, которое выпустило наибольшее число книг (исключая брошюры < 50 страниц)
4. определить автора с самой высокой средней оценкой книг (книги с 50 и более оценками)
5. определить среднее количество обзоров от пользователей, которые поставили больше 48 оценок.


Результаты исследования помогут сформулировать ценностное предложение для нового продукта.

## Описание базы данных

Исследуемая база данных включает в себя 5 таблиц:

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Обзор данных

In [1]:
# импортируем библиотеки необходимые для исследования
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Исследуем данные: выведем 5 первых строк каждой таблицы и общую информацию о ней:

In [2]:
#названия таблицы
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']

#вывод первых 5 строк таблицы и общей информации о ней
for t in tables :
    query = '''SELECT * FROM {}'''.format(t) 
    
    con=engine.connect()
    
    table = pd.io.sql.read_sql(sql=text(query), con = con)
    print('-'* 90)
    print(f'Таблица: {t}')
    display(table.head())
    table.info()

------------------------------------------------------------------------------------------
Таблица: books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
------------------------------------------------------------------------------------------
Таблица: authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
------------------------------------------------------------------------------------------
Таблица: publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
------------------------------------------------------------------------------------------
Таблица: ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
------------------------------------------------------------------------------------------
Таблица: reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


На первый взгляд можно сделать следующие выводы: пропущенных значений в данных нет (количество строк по столбцам одинаковое по каждой таблице).

## Задача 1

Определить количество книг, которые вышли после 1 января 2000 года

In [3]:
query1 = '''
SELECT COUNT(book_id) 
FROM books 
WHERE publication_date > '2000-1-1';
'''

query1 = pd.io.sql.read_sql(sql=text(query1), con = con)
print(f'Количество книг, выпущенных после 1 января 2000 года: {query1.iloc[0, 0]}')

Количество книг, выпущенных после 1 января 2000 года: 819


**Вывод:** из 1000 книг содержащихся в таблице `books` 819 выпущены после 1 января 2000 года, т.е. почти 82% книг исследуемой базы данных относятся к современной литературе.

## Задача 2

Определить количество обзоров и среднюю оценку для каждой книги

In [4]:
query2 = '''
SELECT book_id,
       title,
       COUNT(DISTINCT review_id) AS count_reviews,
       ROUND(AVG(rating), 2) AS avg_rating
FROM books
LEFT JOIN reviews USING (book_id)
LEFT JOIN ratings USING (book_id)
GROUP BY book_id
ORDER BY count_reviews DESC;
'''

query2 = pd.io.sql.read_sql(sql=text(query2), con = con)
display(query2)

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     count_reviews  avg_rating  
0                7        3.66  
1                6        3.98  
2                6        4.21  
3                6        4.41  
4                6        4.08  
..             ...         ...  
995              0        3.67  
996              0        3.00  
997              0        5.00  
998              0        4.00  
999              0        4.00  

[1000 rows x 4 columns]

**Вывод:** Данные выведены в порядке убывания количества пользовательских обзоров, так можно увидеть, что не все книги имеют пользовательские обзоры. Вместе с тем, если вывести данные в порядке возрастания средней оценки книги, то мы увидим что все книги имеют оценки.

In [5]:
display(query2.sort_values(by='avg_rating'))

book_id                                              title  \
651      303                               Harvesting the Heart   
685      371                                              Junky   
659      316                  His Excellency: George Washington   
510      202                                      Drowning Ruth   
497      915  The World Is Flat: A Brief History of the Twen...   
..       ...                                                ...   
661      321                                Homage to Catalonia   
827      699  The Demon-Haunted World: Science as a Candle i...   
592      169  Crucial Conversations: Tools for Talking When ...   
802      642                 The Big Bad Wolf  (Alex Cross  #9)   
726      459                               My Name Is Asher Lev   

     count_reviews  avg_rating  
651              2        1.50  
685              2        2.00  
659              2        2.00  
510              3        2.00  
497              3        2.25  
..             ...         ...  
661              2        5.00  
827              2        5.00  
592              2        5.00  
802              2        5.00  
726              2        5.00  

[1000 rows x 4 columns]

## Задача 3

Определить издательство, которое выпустило наибольшее число книг (исключая брошюры < 50 страниц)

`Вариант 1 (если у нас только 1 издательство с максимальным количеством книг можно использовать LIMIT, так запрос выглядит локоничнее)`

In [6]:

query3 = '''
SELECT publisher,
       COUNT(book_id)
FROM books 
JOIN publishers USING(publisher_id)
WHERE num_pages >= 50
GROUP BY publisher
ORDER BY count DESC
LIMIT 1;
'''

query3 = pd.io.sql.read_sql(sql=text(query3), con = con)
#display(query3)

print(f'Издательство-лидер по количеству выпущенных книг: {query3.iloc[0, 0]}')
print(f'Количество выпущенных книг: {query3.iloc[0, 1]}')

Издательство-лидер по количеству выпущенных книг: Penguin Books
Количество выпущенных книг: 42


**Вывод:** Издательство которые выпустило больше всего книг Penguin Books.

`Вариант 2 (если бы было несколько издательств с максимальным количеством книг то следовало бы использовать MAX)`

In [7]:
query3_1 = '''
WITH

--таблица содержащая название издательства и количество выпущенных им книг
a AS
(SELECT publisher,
        COUNT(book_id) AS count_books
FROM books 
JOIN publishers USING(publisher_id)
WHERE num_pages >= 50
GROUP BY publisher
)

--запрос который выбирает издательство(а) с максимальным количеством выпущенных книг
SELECT publisher,
       count_books
FROM a
WHERE count_books IN (SELECT MAX(count_books)
                      FROM a)
'''

query3_1 = pd.io.sql.read_sql(sql=text(query3_1), con = con)
display(query3_1)

print(f'Издательство-лидер по количеству выпущенных книг: {query3_1.iloc[0, 0]}')
print(f'Количество выпущенных книг: {query3_1.iloc[0, 1]}')

publisher  count_books
0  Penguin Books           42

Издательство-лидер по количеству выпущенных книг: Penguin Books
Количество выпущенных книг: 42


## Задача 4

Определить автора с самой высокой средней оценкой книг (книги с 50 и более оценками)

`Как я поняла задачу:
Определить автора с самой высокой средней оценкой средних оценок его книг. То есть сначала для каждого автора определяем среднюю оценку каждой его книги (ведь у каждой книги может быть несколько оценок) и уже потом находим среднюю оценку всех книг (то есть среднюю оценку средних оценок всех книг 1 автора).
При этом нас интересуют только книги с количеством оценок >=50.`

In [8]:
query4 = '''
WITH
--таблица которая содержит авторов, их книги с количеством оценок >= 50 и среднюю оценку каждой такой книги автора
avg_rat_book AS
(SELECT author_id,
       author,
       book_id,
       avg(rating) AS avg_rat_book,
       count(rating) AS count_ratings
FROM books AS b
JOIN authors AS a USING(author_id)
JOIN ratings AS r USING(book_id)
GROUP BY author_id, author, book_id
HAVING count(rating) >= 50
ORDER BY author_id)

--таблица которая содержит автора и среднюю оценку средних оценок его книг
SELECT a.author,
       round(AVG(a.avg_rat_book), 2) AS avg_rating
FROM avg_rat_book AS a
GROUP BY a.author
ORDER BY avg_rating DESC
LIMIT 1;
'''

query4 = pd.io.sql.read_sql(sql=text(query4), con = con)
#display(query)
print(f'Автор с самой высокой средней оценкой книг: {query4.iloc[0, 0]}')
print(f'Средняя оценка книг: {query4.iloc[0, 1]}')

Автор с самой высокой средней оценкой книг: J.K. Rowling/Mary GrandPré
Средняя оценка книг: 4.28


**Вывод:** Автор с самой высокой средней оценкой книг: J.K. Rowling/Mary GrandPré

## Задача 5

Определить среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [9]:
query5 = '''
SELECT ROUND(AVG(count_reviews), 2)
FROM (
--таблица которая содержит пользователей, количество оценок поставленных ими 
--и количество обзоров каждого пользователя
    SELECT username,
           COUNT(DISTINCT rating_id) AS count_ratings,
           COUNT(DISTINCT review_id) AS count_reviews
    FROM ratings
    JOIN reviews USING(username)
    GROUP BY username
    HAVING COUNT(DISTINCT rating_id) > 48) AS a
'''

query5 = pd.io.sql.read_sql(sql=text(query5), con = con)
#display(query5)
print(f'Среднее количество обзоров от пользователей которые поставили больше \
48 оценок: {query5.iloc[0, 0]} обзоров')

Среднее количество обзоров от пользователей которые поставили больше 48 оценок: 24.0 обзоров


**Вывод:** Пользователи, которые поставили больше 48 оценок, пишут в среднем около 24 обзоров на книги.

# Результаты исследования

Результатом исследования стали ответы на все поставленные перед исследованием вопросы, их можно собрать в единую характеристику имеющейся базы данных книг:
1. **819 книг** из 1000 имеющихся **были выпущенных после 1 января 2000 года**
2. была сформулирована таблица (задача 2) с количеством обзоров и средней оценкой для каждой книги
3. **Penguin Books** - издательство, которое выпустило наибольшее число книг (исключая брошюры < 50 страниц)
4. автор с самой высокой средней оценкой книг (среди книг с 50 и более оценками) - **J.K. Rowling/Mary GrandPré**
5. пользователи, которые поставили больше 48 оценок, пишут **в среднем около 24 обзоров** на книги.